### PR TITLE
feat(mechanics): Allow escorts to hold fire

### PR DESCRIPTION
**Pull request by Koranir**
**Feature:** This PR implements the feature request detailed and discussed in issue #7543 , #645, and many more.

## Feature Details
The 'X' key now stops escorts from using offensive weapons.

'Hold position' has been renamed to 'Stay in position'.

## UI Screenshots
image

image

https://user-images.githubusercontent.com/101683475/200434434-d22bdd9b-5140-4d1a-b825-5a9b3f5e3642.mp4

## Testing Done
Went into battle with my fleet and pressed H again and again. Worked fine.

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

**Comment by Koranir** [comment]
Took 6 years, but here it is.

File: source/AI.cpp [hunk 1]
```diff
@@ -3006,10 +3011,15 @@ void AI::AutoFire(const Ship &ship, FireCommand &command, bool secondary) const
...
+		if(it != orders.end())
 		{
-			disabledOverride = (it->second.type == Orders::FINISH_OFF);
-			friendlyOverride = disabledOverride | (it->second.type == Orders::ATTACK);
+			if(it->second.target.lock() == currentTarget)
+			{
+				disabledOverride = (it->second.type == Orders::FINISH_OFF);
+				friendlyOverride = disabledOverride | (it->second.type == Orders::ATTACK);
+			}
+			if(it->second.type == Orders::HOLD_FIRE)
```
Comment (Hurleveur) [inline source/AI.cpp:3021]: If we use this as a return it should be before the lines just above
That said, I think that an order to fire should overwrite the default hold fire order, just for this target
So just make it an else if?

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Looks good I dont have much issue with it
Somebody should test it ofc but since there is a vid there shouldnt be an issue

**Review by Hurleveur — APPROVED** [review]

File: source/AI.cpp [hunk 2]
```diff
@@ -3006,10 +3011,15 @@ void AI::AutoFire(const Ship &ship, FireCommand &command, bool secondary) const
...
 		auto it = orders.find(&ship);
-		if(it != orders.end() && it->second.target.lock() == currentTarget)
+		if(it != orders.end())
 		{
-			disabledOverride = (it->second.type == Orders::FINISH_OFF);
-			friendlyOverride = disabledOverride | (it->second.type == Orders::ATTACK);
+			if(it->second.target.lock() == currentTarget)
+			{
+				disabledOverride = (it->second.type == Orders::FINISH_OFF);
+				friendlyOverride = disabledOverride | (it->second.type == Orders::ATTACK);
```
Comment (quyykk) [inline source/AI.cpp:3019]:
```suggestion
				friendlyOverride = disabledOverride || (it->second.type == Orders::ATTACK);
```

while we're at it.

**Review by quyykk — COMMENTED** [review]
I'm not too sure if I'm a fan of changing the (default) keybind for "hold position" and basically reseting the existing keybind that user's have.

**Comment by Koranir** [comment]
> I'm not too sure if I'm a fan of changing the (default) keybind for "hold position" and basically reseting the existing keybind that user's have.

It should be fine in my opinion, and I would rather the default keybind for holding fire to be closer to the left hand.

**Comment by Hurleveur** [comment]
I think it'd be best to not change it rn and instead do an overhaul of the keys later on if its required, as that is pretty much beyond the scope of the PR.

**Comment by Koranir** [comment]
> I think it'd be best to not change it rn and instead do an overhaul of the keys later on if its required, as that is pretty much beyond the scope of the PR. (change that, this will be merged and we worry about hotkeys in another PR)

I thought over it and decided to change the default key to x.

**Comment by mike-f1** [comment]
Nice to have!

File: keys.txt [hunk 3]
```diff
@@ -22,5 +22,6 @@
 "Toggle fast-forward" 1073741881
 "Fleet: Fight my target" 102
 "Fleet: Gather around me" 103
-"Fleet: Hold position" 104
+"Fleet: Hold fire" 120
+"Fleet: Stay in position" 104
```
Comment (warp-core) [inline keys.txt:26]: Why rename `Hold position`?
`Hold fire` seems like it should go near `Fight my target`.
```suggestion
"Fleet: Hold fire" 120
"Fleet: Gather around me" 103
"Fleet: Hold position" 104
```

File: source/Command.h [hunk 4]
```diff
@@ -61,6 +61,7 @@ class Command {
 	static const Command FIGHT;
 	static const Command GATHER;
 	static const Command HOLD;
+	static const Command STAY;
```
Comment (warp-core) [inline source/Command.h:64]: Just call them `HOLD_POSITION` and `HOLD_FIRE`.

**Review by warp-core — CHANGES_REQUESTED** [review]

File: keys.txt (hunk 3 above)
Comment (Hurleveur) [reply keys.txt:26]: This used to change the value for hold position so I assume thats why

File: source/Command.cpp [hunk 5]
```diff
@@ -70,11 +70,12 @@ const Command Command::FULLSCREEN(1uL << 21, "Toggle fullscreen");
 const Command Command::FASTFORWARD(1uL << 22, "Toggle fast-forward");
 const Command Command::FIGHT(1uL << 23, "Fleet: Fight my target");
 const Command Command::GATHER(1uL << 24, "Fleet: Gather around me");
-const Command Command::HOLD(1uL << 25, "Fleet: Hold position");
-const Command Command::AMMO(1uL << 26, "Fleet: Toggle ammo usage");
-const Command Command::WAIT(1uL << 27, "");
-const Command Command::STOP(1ul << 28, "");
-const Command Command::SHIFT(1uL << 29, "");
+const Command Command::HOLD(1uL << 25, "Fleet: Hold fire");
+const Command Command::STAY(1uL << 26, "Fleet: Stay in position");
```
Comment (Amazinite) [inline source/Command.cpp:74]: Not a fan of the rename.

File: keys.txt [hunk 6]
```diff
@@ -21,6 +21,7 @@
 "Toggle fullscreen" 1073741892
 "Toggle fast-forward" 1073741881
 "Fleet: Fight my target" 102
+"Fleet: Hold fire" 120
```
Comment (warp-core) [inline keys.txt:24]: I would suggest using the 'y' key for this. The other fleet commands are grouped reasonably close: D, F, G, and H, with toggle ammo usage being U.
```suggestion
"Fleet: Hold fire" 121
```

**Review by warp-core — COMMENTED** [review]
Ships that are holding fire seem to act really strangely. I think they're still targeting and pursuing hostiles as normal, but just not shooting at them?

And, then, if I tell them to gather around me, probably something I want them to do if there are hostiles around that I want them not to fire on, since that leaves them vulnerable themselves, they start firing again?
Same if I tell them to hold position.

And then, if I tell them to hold fire, they stop gathering around me or holding position.

These are not things that should be mutually exclusive.

**Comment by Hurleveur** [comment]
@Koranir I think that this feature would be a huge boon to have, and it's a shame for it to not make it in for a few relatively small things
can you maybe take a look at them? if not I can just do a PR real quick, or even PR it myself to take care of things requested and more

**Comment by ziproot** [comment]
@Hurleveur Koranir says you can take over if you want. 

[Source](https://discord.com/channels/251118043411775489/285540320823869441/1166210167503855646)

**Comment by bene-dictator** [comment]
Shall I close this?

**Pull request closed without merging**